sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/BlobDecryptionPolicy.java

```diff
@@ -163,30 +146,21 @@ public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineN
                 }
             });
         } else {
-            // Issue a get properties call to determine how much we will need to expand the range.
-            // Apply the request conditions from the download to the getProperties.
-            BlobRequestConditions rc = extractRequestConditionsFromRequest(requestHeaders);
-            return this.blobClient.getPropertiesWithResponse(rc).flatMap(response -> {
-                EncryptionData encryptionData = EncryptionData.getAndValidateEncryptionData(
-                    response.getValue().getMetadata().get(CryptographyConstants.ENCRYPTION_DATA_KEY),
-                    requiresEncryption);
-
-                // Apply the etag from the getProperties to the download to ensure consistency
-                String etag = response.getValue().getETag();
-                requestHeaders.set("ETag", etag);
-
-                // If there is no encryption data, then the blob is not encrypted. Pass the request through.
-                if (encryptionData == null) {
-                    return next.process();
-                }
+            // If it was a ranged request, we would have already called get properties and set encryption data.
+            // Since there is no encryption data, the request is not encrypted
+            if (context.getData(CryptographyConstants.ENCRYPTION_DATA_KEY).isEmpty()) {
+                return next.process();
+            }
```

Comment:
I'd validate that response.metadata doesn't have encryption_data_key in addition to that and throw illegalstateexception.
We don't have compiler assisted way to assure that the assumption here holds, so we should throw a validation in case high level apis and this policy misalign.

Reply:
The worry I have is that we have implicit coupling between what's happening in clients with the policy code. If somebody modifies them in the future without understanding of those couplings they're going to have hard time understanding the error they observe.
So if they attempt to read encrypted range without calling to get props (because they develop new api for instance) they're going to have bad time understanding why they observe decryption error. If we throw them validation error they'll know faster.

On the other hand, now that I think about this. if we just add validation here with current shape of logic the getProperties might hit it and fail (because it would have metadata but nothing in context, or maybe I totally got lost in the logic flow of the policy).

Perhaps it would be good to refactor the policy to the point where this is easier to track for naive reader? (I.e. the fact that range download and get properties fall into same branch is disturbing, working but disturbing).
```
if (isDownload()) {
   throwOnContextMetadataMismatch();
   if (encrypted) {
      decrypt();
   } else {
     next();
   }
} else if (isRangeDonwload()) {
   throwOnContextMetadataMismatch();
   if (encrypted) {
      decrypt();
   } else {
     next();
   }
} else {
  // all other passthrough
  next()
}

```

Reply:
I added the validation and refactored some of the checks to be more readable. Let me know if that helps or needs more work. I didn't rearrange the structure because I think the way the reactor flows are linked up it would make the method a lot more verbose to do it exactly as you proposed.